### PR TITLE
Upgrade requests lib : axios and server proxy

### PR DESCRIPTION
**Pull request by chrichri17**
This PR update the `requests` library used to make axios requests

## What should you do now ?
- Create an `.env` file in `quizoot/frontend` with the variable `VITE_QUIZOOT_SERVER_PORT=8000`(or another port your server is listening on)
- To make a request to the quizoot api, just import `makeRequest` from `lib/requests`

## What's next ?
- CSRF config to be able to make post requests and remove `@csrf_exempt` from backend

### CODE REVIEW

**Comment by chrichri17** [comment]
Current dependencies on/for this PR:
* main
  * **PR #49** <a href="https://app.graphite.dev/github/pr/YacineSteeve/quizoot/49" target="_blank">Graphite</a> 
    * **PR #50** <a href="https://app.graphite.dev/github/pr/YacineSteeve/quizoot/50" target="_blank">Graphite</a> 
      * **PR #51** <a href="https://app.graphite.dev/github/pr/YacineSteeve/quizoot/51" target="_blank">Graphite</a> 
        * **PR #53** <a href="https://app.graphite.dev/github/pr/YacineSteeve/quizoot/53" target="_blank">Graphite</a>  👈
        * **PR #52** <a href="https://app.graphite.dev/github/pr/YacineSteeve/quizoot/52" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/YacineSteeve/quizoot/53?utm_source=stack-comment).

File: frontend/src/lib/requests.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+export async function makeRequest<TData extends ApiResponse>(
+    config: AxiosRequestConfig
+): Promise<TData> {
+    try {
+        const response = await quizootApi(config);
+        return response.data;
+    } catch (err) {
+        if (axios.isAxiosError(err)) {
+            // TODO: display
```
Comment (chrichri17) [inline frontend/src/lib/requests.ts:28]: For axios error, we can probably do something better with the error message

File: frontend/vite.config.ts [hunk 2]
```diff
@@ -1,15 +1,28 @@
...
+
+    return {
+        plugins: [vue(), vueJsx()],
+        resolve: {
+            alias: {
+                '@': fileURLToPath(new URL('./src', import.meta.url)),
+            },
+        },
+        server: {
+            proxy: {
```
Comment (chrichri17) [inline frontend/vite.config.ts:20]: every requests should starts with `/api`. 

`fetch('/api/quizzes')` will be redirected to `http://localhost:8000/api/quizzes`.

Comment (YacineSteeve) [reply frontend/vite.config.ts:20]: Can't we do it like this ?
```typescript
proxy: {
    '/': {
        target: `http://localhost:${PORT}/api`,
        changeOrigin: true,
     },
}
```
So as `fetch('/quizzes')` would be redirected to `http://localhost:8000/api/quizzes`

Comment (chrichri17) [reply frontend/vite.config.ts:20]: No ! Because we want to route frontend traffic to backend.

`fetch('/api/quizzes')` is a request with the full path `http://localhost:FRONTEND_PORT/api/quizzes` and that will be proxied to `http://localhost:BACKEND_PORT/api/quizzes`.

We cannot use `/` because frontend will route `http://localhost:FRONTEND_PORT/quizzes` to `http://localhost:BACKEND_PORT/api/quizzes`. But you know that `http://localhost:FRONTEND_PORT/quizzes` is the frontend VIEW that list all the quizzes (basically it is the HOME component)...  you don't want to re-route that. It will be like telling that ALL the frontend traffic is routed the api.


Basically, we only route traffic from `FRONTEND/api` to `BACKEND/api` since frontend doesn't implement the services.

**Comment by YacineSteeve** [comment]
> ## What should you do now ?
> * Create an `.env` file in `quizoot/frontend` with the variable `VITE_QUIZOOT_SERVER_PORT=8000`(or another port your server is listening on)

What if you provide a `.env.example` as done for the `backend` ?

**Comment by chrichri17** [comment]
> > ## What should you do now ?
> > 
> > * Create an `.env` file in `quizoot/frontend` with the variable `VITE_QUIZOOT_SERVER_PORT=8000`(or another port your server is listening on)
> 
> What if you provide a `.env.example` as done for the `backend` ?

Ok

File: frontend/src/lib/requests.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+export type ApiResponse =
+    | number
+    | string
+    | boolean
+    | null
+    | ApiResponse[]
+    | { [x: string]: ApiResponse };
+
+export async function makeRequest<TData extends ApiResponse>(
+    config: AxiosRequestConfig
```
Comment (YacineSteeve) [inline frontend/src/lib/requests.ts:21]: Isn't `request` a better name than `config` ?

Comment (chrichri17) [reply frontend/src/lib/requests.ts:21]: I basically used the same prototype used by axios itself:

```typescript
<any, AxiosResponse<any, any>, any>(config: AxiosRequestConfig<any>) => Promise<AxiosResponse<any, any>>

// The whole interface is as follow in frontend/node_modules/axios/index.d.ts:
export interface AxiosInstance extends Axios {
  <T = any, R = AxiosResponse<T>, D = any>(config: AxiosRequestConfig<D>): Promise<R>;
  <T = any, R = AxiosResponse<T>, D = any>(url: string, config?: AxiosRequestConfig<D>): Promise<R>;

  defaults: Omit<AxiosDefaults, 'headers'> & {
    headers: HeadersDefaults & {
      [key: string]: AxiosHeaderValue
    }
  };
}
```

And as the name `AxiosRequestConfig` is suggesting, it is a configuration. Does that convince you ? We can name it `requestConfig` if you want

File: frontend/vite.config.ts (hunk 2 above)
Comment (YacineSteeve) [reply frontend/vite.config.ts:20]: Oh I see now. A big misundersting of how the proxy actually works: `Any requests that request path starts with that key will be proxied to that specified target` (from [Vite Doc](https://vitejs.dev/config/server-options.html#server-proxy)).
Thanks !

File: frontend/src/lib/requests.ts (hunk 3 above)
Comment (YacineSteeve) [reply frontend/src/lib/requests.ts:21]: `requestConfig` is Ok

File: frontend/src/lib/requests.ts (hunk 1 above)
Comment (YacineSteeve) [reply frontend/src/lib/requests.ts:28]: I don't know for now. "Maybe" you can find any idea about that in this [article](https://dev.to/mperon/axios-error-handling-like-a-boss-333d)

**Review by YacineSteeve — APPROVED** [review]
